[WebAuthn] Add getPublicKey/Algorithm to AuthenticatorAttestationResponse

pascoej opened the pull request:
#### 9e683e62ce3cd50823af5214c5afa4bf08a13ec6
<pre>
[WebAuthn] Add getPublicKey/Algorithm to AuthenticatorAttestationResponse
<a href="https://bugs.webkit.org/show_bug.cgi?id=241878">https://bugs.webkit.org/show_bug.cgi?id=241878</a>
rdar://90281745

Reviewed by Brent Fulgham.

This patch implements the getPublicKeyAlgorithm and getPublicKey convenience
methods described in <a href="https://www.w3.org/TR/webauthn-2/#sctn-public-key-easy">https://www.w3.org/TR/webauthn-2/#sctn-public-key-easy</a>

These allow RPs to import keys in a more widely supported format than the
COSE key returned in the attestationObject. getPublicKey() is only implemented
for ES256 keys, which is the first default pubKeyCredParams.

Tested using wpt/webauthn/createcredential-getpublickey.https.html

* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
(WebCore::coseKeyForAttestationObject):
(WebCore::AuthenticatorAttestationResponse::getPublicKeyAlgorithm const):
(WebCore::AuthenticatorAttestationResponse::getPublicKey const):
* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.h:
* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.idl:
* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.cpp:
(WebCore::encodeRawPublicKey):
* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.h:
* Source/WebCore/Modules/webauthn/fido/Pin.cpp:
(fido::pin::encodeRawPublicKey): Deleted.
* Source/WebCore/Modules/webauthn/fido/Pin.h:
* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-local.https-expected.txt
* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-local.https.html

Canonical link: <a href="https://commits.webkit.org/251844@main">https://commits.webkit.org/251844@main</a>
</pre>

brentfulgham reviewed (commented):
I don't see a new test for these new methods. Will they be exercised in an upcoming change?

I do see that the change in Pin is tested, so the move of the code is validated.

brentfulgham reviewed (approved):
Thanks for adding the tests! r=me

pascoej committed:
[WebAuthn] Add getPublicKey/Algorithm to AuthenticatorAttestationResponse
https://bugs.webkit.org/show_bug.cgi?id=241878
rdar://90281745

Reviewed by Brent Fulgham.

This patch implements the getPublicKeyAlgorithm and getPublicKey convenience
methods described in https://www.w3.org/TR/webauthn-2/#sctn-public-key-easy

These allow RPs to import keys in a more widely supported format than the
COSE key returned in the attestationObject. getPublicKey() is only implemented
for ES256 keys, which is the first default pubKeyCredParams.

Tested using wpt/webauthn/createcredential-getpublickey.https.html

* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
(WebCore::coseKeyForAttestationObject):
(WebCore::AuthenticatorAttestationResponse::getPublicKeyAlgorithm const):
(WebCore::AuthenticatorAttestationResponse::getPublicKey const):
* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.h:
* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.idl:
* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.cpp:
(WebCore::encodeRawPublicKey):
* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.h:
* Source/WebCore/Modules/webauthn/fido/Pin.cpp:
(fido::pin::encodeRawPublicKey): Deleted.
* Source/WebCore/Modules/webauthn/fido/Pin.h:
* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-local.https-expected.txt
* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-local.https.html

Canonical link: https://commits.webkit.org/251844@main
